woocommerce/wc-admin
# Improve the pre-commit hook setup for wc-admin

Our current pre-commit hook setup is very slow and almost never useful. The goal of this setup change is to make the pre-commit checks as useful to you as possible and as quick as possible. 

The biggest highlights of these changes are that CSS and JS are autoformatted and added to your PR, and also that Jest will run just tests related to changes and fail early if one fails instead of running the whole suite redundantly.

### What it does

On pre-commit the lint-staged setup now does the following:

1. If CSS has changed it will lint and fix the CSS automatically. Anything fixed will be auto added to the commit thanks to `lint-staged`'s ability to auto stage any changes made by formatters.

2. If any changes were made in packages or client, it will do the following:
  2a. Format the JS (prettier) (if anything was changed it will auto add it as above)
  2b. Lint the JS (eslint) 
  2c. Run tests but only tests related to your changes and bail early if one fails.

3. Lint the PHP

On pre-push the only thing that happens now is a GPL reminder text and a check you're not pushing to `main`. The `main` check is redundant due to branch protection in Github, but its a low cost to leave this here so I'm not changing it.

### How to test

To test this feature, you can create a dummy branch from this branch, e.g. `git checkout -b my-new-branch`. Add some changes to JS, CSS or PHP and see how it behaves in each scenario. You could test:

1. Adding some JS that is formatted incorrectly.
2. Adding some JS that will have lint errors (like console.log statements)
3. Adding some PHP code with lint errors

Once the changes are made, make a commit and watch the husky/lint-staged workflow do its magic ✨ . A final extra credit step would be performing a git push and noting the presence of the GPL reminder.
